micrometer-java21/src/main/java/io/micrometer/java21/instrument/binder/jdk/VirtualThreadMetrics.java

```diff
@@ -0,0 +1,107 @@
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.java21.instrument.binder.jdk;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.binder.MeterBinder;
+import jdk.jfr.consumer.RecordingStream;
+
+import java.io.Closeable;
+import java.time.Duration;
+import java.util.Objects;
+
+import static java.util.Collections.emptyList;
+
+/**
+ * Instrumentation support for Virtual Threads, see:
+ * https://openjdk.org/jeps/425#JDK-Flight-Recorder-JFR
+ *
+ * @author Artyom Gabeev
+ * @since 1.14.0
+ */
+public class VirtualThreadMetrics implements MeterBinder, Closeable {
+
+    private static final String PINNED_EVENT = "jdk.VirtualThreadPinned";
+
+    private static final String SUBMIT_FAILED_EVENT = "jdk.VirtualThreadSubmitFailed";
+
+    private final RecordingStream recordingStream;
+
+    private final Iterable<Tag> tags;
+
+    public VirtualThreadMetrics() {
+        this(new RecordingConfig(), emptyList());
+    }
+
+    public VirtualThreadMetrics(Iterable<Tag> tags) {
+        this(new RecordingConfig(), tags);
+    }
+
+    private VirtualThreadMetrics(RecordingConfig config, Iterable<Tag> tags) {
+        this.recordingStream = createRecordingStream(config);
+        this.tags = tags;
+    }
+
+    @Override
+    public void bindTo(MeterRegistry registry) {
+        Timer pinnedTimer = Timer.builder("jvm.threads.virtual.pinned")
+            .description("The duration while the virtual thread was pinned without releasing its platform thread")
+            .tags(tags)
+            .register(registry);
+
+        Counter submitFailedCounter = Counter.builder("jvm.threads.virtual.submit.failed")
+            .description("The number of events when starting or unparking a virtual thread failed")
+            .tags(tags)
+            .register(registry);
+
+        recordingStream.onEvent(PINNED_EVENT, event -> pinnedTimer.record(event.getDuration()));
+        recordingStream.onEvent(SUBMIT_FAILED_EVENT, event -> submitFailedCounter.increment());
+    }
+
+    private RecordingStream createRecordingStream(RecordingConfig config) {
+        RecordingStream recordingStream = new RecordingStream();
+        recordingStream.enable(PINNED_EVENT).withThreshold(config.pinnedThreshold);
+        recordingStream.enable(SUBMIT_FAILED_EVENT);
+        recordingStream.setMaxAge(config.maxAge);
+        recordingStream.setMaxSize(config.maxSizeBytes);
+        recordingStream.startAsync();
```

Comment:
It's probably not a big deal and in most cases bind will be called very soon after this class is instantiated, but it would make more sense to start the stream only when the meters are going to be bound (when `bindTo` is called). To do that, though, we would have to bring back the `started` flag and worry about concurrent calls to `bindTo` to be safest. Maybe it's not worth it to avoid a stream being created and started even though `bindTo` may be called later (or never).

If we had more instrumentation that used JFR, it would be a good idea to have some centralized infrastructure around it so we aren't making many RecordingStream instances where one could be re-used, but it's not apparently worth doing that when we only have one instrumentation that uses it, I think.

Reply:
I think once we add more instrumentation to JFR, we can create a component that manages starting the recording stream and the registration in a thread-safe way.

Reply:
I think JFR may be interesting source for observability, lets park it as is until we have another metrics derived from JFR and rethink record stream lifecycle management.

Reply:
Thanks to JFR being an implementation detail of this instrumentation, we should hopefully be able to accomplish that in a way that causes the least disruption later.